File: thirdai_platform_py/llm_cache_job/cache.py
import os
from abc import ABC, abstractmethod
from logging import Logger
from typing import Any, Dict, List, Optional, Set

from thirdai import neural_db_v2 as ndb
from thirdai.neural_db_v2.chunk_stores import constraints


class Cache(ABC):
    @abstractmethod
    def suggestions(self, model_id: str, query: str) -> List[Dict[str, Any]]:
        raise NotImplemented

    @abstractmethod
    def query(self, model_id: str, query: str) -> Optional[Dict[str, Any]]:
        raise NotImplemented

    @abstractmethod
    def insert(self, model_id: str, query: str, llm_res: str) -> None:
        raise NotImplemented

    @abstractmethod
    def invalidate(self, model_id: str) -> None:
        raise NotImplemented


def token_similarity(query_tokens: Set[str], cached_query: str) -> float:
    overlap = len(query_tokens.intersection(cached_query.split()))
    return overlap / len(query_tokens)


class NDBSemanticCache(Cache):
    def __init__(self, logger: Logger):
        self.logger = logger
        self.cache_ndb_path = os.path.join(
            os.getenv("MODEL_BAZAAR_DIR"), "llm_cache.ndb"
        )
        self.logger.info(f"cache ndb at {self.cache_ndb_path}")
        if os.path.exists(self.cache_ndb_path):
            try:
                self.db = ndb.NeuralDB.load(self.cache_ndb_path)
                self.logger.info("Loaded existing cache model from disk.")
            except Exception as e:
                self.logger.error("Failed to load cache model", exc_info=True)
                raise e
        else:
            try:
                self.db = ndb.NeuralDB(save_path=self.cache_ndb_path)
                self.logger.info("Initialized new cache model.")
            except Exception as e:
                self.logger.error("Failed to initialize new cache model", exc_info=True)
                raise e
        self.threshold = float(os.getenv("LLM_CACHE_THRESHOLD", "0.95"))
        self.logger.info(f"Cache threshold set to {self.threshold}")

    def suggestions(self, model_id: str, query: str) -> List[Dict[str, Any]]:
        self.logger.info(
            f"Fetching suggestions for model_id={model_id}, query='{query}'"
        )
        if self.db.retriever.retriever.size() == 0:
            self.logger.info("No suggestions found; cache is empty.")
            return []

        results = self.db.search(
            query=query,
            top_k=5,
            constraints={"model_id": constraints.EqualTo(model_id)},
        )

        unique_suggestions = set()
        suggestions = []
        for res, _ in results:
            if res.text in unique_suggestions:
                continue
            suggestions.append({"query": res.text, "query_id": res.chunk_id})
            unique_suggestions.add(res.text)

        return suggestions

    def query(self, model_id: str, query: str) -> Optional[str]:
        self.logger.info(
            f"Executing cache query for model_id={model_id}, query='{query}'"
        )
        if self.db.retriever.retriever.size() == 0:
            self.logger.info("Cache is empty;")
            return None

        results = self.db.search(
            query=query,
            top_k=5,
            constraints={"model_id": constraints.EqualTo(model_id)},
        )

        query_tokens = set(query.split())
        reranked = sorted(
            [(res[0], token_similarity(query_tokens, res[0].text)) for res in results],
            key=lambda x: x[1],
            reverse=True,
        )

        if len(reranked) > 0 and reranked[0][1] > self.threshold:
            self.logger.info(
                f"Cache hit with similarity {reranked[0][1]} for query '{query}'"
            )
            return {
                "query": reranked[0][0].text,
                "query_id": reranked[0][0].chunk_id,
                "llm_res": reranked[0][0].metadata["llm_res"],
            }

        self.logger.info("Cache miss or similarity below threshold.")
        return None

    def insert(self, model_id: str, query: str, llm_res: str) -> None:
        self.logger.info(f"Inserting query into cache for model_id={model_id}")
        self.db.insert(
            [
                ndb.InMemoryText(
                    document_name="",
                    text=[query],
                    doc_metadata={"model_id": model_id, "llm_res": llm_res},
                )
            ]
        )

    def invalidate(self, model_id: str) -> None:
        self.logger.info(f"Invalidating cache entries for model_id={model_id}")
        ids = self.db.chunk_store.filter_chunk_ids(
            constraints={"model_id": constraints.EqualTo(model_id)}
        )

        self.db.delete(list(ids))

File: thirdai_platform/deployment/cache.go
package deployment

import (
	"fmt"
	"log/slog"
	"path/filepath"
	"slices"
	"sort"
	"strconv"
	"strings"
	"thirdai_platform/search/ndb"
)

type LLMCache struct {
	Ndb       ndb.NeuralDB
	Threshold float64
}

const CacheScoreThreshold = 0.95

func NewLLMCache(modelBazaarDir, modelId string) (*LLMCache, error) {
	cachePath := filepath.Join(modelBazaarDir, "models", modelId, "llm_cache", "llm_cache.ndb")
	ndb, err := ndb.New(cachePath)
	if err != nil {
		return nil, fmt.Errorf("unable to construct LLM Cache: %v", err)
	}

	return &LLMCache{Ndb: ndb, Threshold: CacheScoreThreshold}, nil
}

func (c *LLMCache) Close() {
	c.Ndb.Free()
}

func (c *LLMCache) Suggestions(query string) ([]string, error) {
	slog.Info("fetching cache suggestions", "query", query)

	chunks, err := c.Ndb.Query(query, 5, nil)
	if err != nil {
		return []string{}, fmt.Errorf("ndb query error: %v", err)
	}

	if len(chunks) == 0 {
		slog.Info("cache is empty")
		return []string{}, nil
	}

	seen := make(map[string]struct{})
	suggestions := []string{}

	for _, chunk := range chunks {
		str := chunk.Text
		if _, exists := seen[str]; !exists {
			seen[str] = struct{}{}
			suggestions = append(suggestions, str)
		}
	}

	return suggestions, nil
}

func tokenSimilarity(queryTokens map[string]struct{}, cachedQuery string) float64 {
	if len(queryTokens) == 0 {
		return 0.0
	}

	cachedTokens := strings.Fields(cachedQuery)
	overlap := 0
	for _, token := range cachedTokens {
		if _, exists := queryTokens[token]; exists {
			overlap++
		}
	}

	return float64(overlap) / float64(len(queryTokens))
}

func getTopChunkSimilarity(query string, chunks []ndb.Chunk) (ndb.Chunk, float64) {
	queryTokens := make(map[string]struct{})
	for _, token := range strings.Fields(query) {
		queryTokens[token] = struct{}{}
	}

	sort.Slice(chunks, func(i, j int) bool {
		return tokenSimilarity(queryTokens, chunks[i].Text) > tokenSimilarity(queryTokens, chunks[j].Text)
	})

	topChunk := chunks[0]
	topSimilarity := tokenSimilarity(queryTokens, topChunk.Text)
	return topChunk, topSimilarity
}

func referenceIdsToString(referenceIds []uint64) string {
	referenceIdString := make([]string, len(referenceIds))
	for i, num := range referenceIds {
		referenceIdString[i] = strconv.FormatUint(num, 10)
	}
	return strings.Join(referenceIdString, " ")
}

func referenceIdsFromString(referenceIdString string) ([]uint64, error) {
	parts := strings.Split(referenceIdString, " ")
	referenceIds := make([]uint64, len(parts))

	for i, part := range parts {
		num, err := strconv.ParseUint(part, 10, 64)
		if err != nil {
			return []uint64{}, fmt.Errorf("could not parse reference ids from metadata %s", referenceIdString)
		}
		referenceIds[i] = num
	}
	return referenceIds, nil
}

func getChunkMetadata(chunk ndb.Chunk) (string, []uint64, error) {
	llmResUncasted, ok := chunk.Metadata["llm_res"]
	if !ok {
		return "", []uint64{}, fmt.Errorf("llm_res metadata value not found")
	}

	llmRes, ok := llmResUncasted.(string)
	if !ok {
		return "", []uint64{}, fmt.Errorf("llm_res metadata value not of string type")
	}

	referenceIDsUncasted, ok := chunk.Metadata["reference_ids"]
	if !ok {
		return "", []uint64{}, fmt.Errorf("llm_res metadata value not found")
	}

	referenceIdsString, ok := referenceIDsUncasted.(string)
	if !ok {
		return "", []uint64{}, fmt.Errorf("llm_res metadata value not of string type")
	}

	referenceIds, err := referenceIdsFromString(referenceIdsString)
	if err != nil {
		return "", []uint64{}, err
	}

	return llmRes, referenceIds, nil
}

func (c *LLMCache) Query(query string, expectedReferenceIds []uint64) (string, error) {
	slog.Info("executing cache request", "query", query)

	chunks, err := c.Ndb.Query(query, 5, nil)
	if err != nil {
		return "", fmt.Errorf("ndb query error: %v", err)
	}

	if len(chunks) == 0 {
		slog.Info("cache returned no results")
		return "", nil
	}

	topChunk, topSimilarity := getTopChunkSimilarity(query, chunks)

	if topSimilarity < c.Threshold {
		slog.Info("top chunk similarity below threshold", "similarity", topSimilarity, "threshold", c.Threshold)
		return "", nil
	}

	llmRes, actualReferenceIds, err := getChunkMetadata(topChunk)
	if err != nil {
		return "", fmt.Errorf("error reading cache chunk metadata: %v", err)
	}

	// if the references match from the original query stored in the cache, we
	// can be pretty certain the llm response is still valid, thus return it
	slices.Sort(expectedReferenceIds)
	slices.Sort(actualReferenceIds)
	if slices.Equal(expectedReferenceIds, actualReferenceIds) {
		return llmRes, nil
	}

	// if the references have changed for the same query, delete it from the cache
	// since the underlying neuraldb has changed and the response might not be valid
	if query == topChunk.Text {
		err := c.Ndb.Delete(query, false)
		if err != nil {
			return "", fmt.Errorf("failed to delete cache entry: %v", err)
		}
	}

	// since the underlying references have changed, we don't return any response here
	return "", nil
}

func (c *LLMCache) Insert(query, llmRes string, referenceIds []uint64) error {
	slog.Info("inserting to cache", "query", query, "llm_res", llmRes)

	err := c.Ndb.Insert(
		"cache_query", query, // use the query as the docId so we can easily delete
		[]string{query},
		[]map[string]interface{}{{"llm_res": llmRes, "reference_ids": referenceIdsToString(referenceIds)}},
		nil)

	if err != nil {
		return fmt.Errorf("failed insertion to cache")
	}

	return nil
}
